### ISSUE TITLE
RFC Checker module, rebooted.

### ISSUE THREAD

**Issue by dset0x** [inveniosoftware/invenio#3345]
Efforts to rewrite the `Checker` module (previously `BibCheck`) have recently been resumed by myself, @Dziolas and @errikos, with guidance from @jalavik and @kaplun. This RFC is meant to replace #2488 and expand on https://github.com/inveniosoftware/invenio/issues/1889?#issuecomment-60728222 .

Due to some new requirements, we are killing some existing code and redesigning the module.

The `Checker` is a framework for running configurable functions meant for, while not being limited to,

* Performing amendments in records
* Composition and dispatching of reports, issues, statistics, etc

Here I will present an overview of definitions and components, without digging into implementation details too much, in order to underline points that we found to be important.

#### Check:
A programmer-provided function that lives in a file. Essentially an improved version of `BibCheck`'s checks. It is capable of

* Requesting modifications to records
* Writing to logs
* Providing parallelization hints (For performance reasons and given that we expect `checks` to be record-centric, a check may declare which records it intends to make changes to as well as which fields it may edit on these records)

`Checks` are given fixtures, such as:
* `all_recids` (all record ids enabled for this run)
* `cfg_args` (func args for this check)
* `record` (a single record)
* `perform_request_search`, etc.
This allows us to optionally wrap these functions appropriately so that that, for instance, we know which records the check has accessed.

#### Check configuration:
Configuration that corresponds to a single `check`, very similar to the one currently present in `BibCheck`. However, it is now stored in the database.

#### Rule:
A pointer to a `check` accompanied by a relevant `check configuration`.

###  Components
Conceptually, the module's architecture will be split in four components. Here they are presented in their usual execution order.

#### Managers:

A user- and admin-facing interfaces for

* Creating, Deleting, Modifying
* Starting, Stopping
* Scheduling, Unscheduling
* Viewing
* Monitoring

`rules`.

* All managers must be capable of performing _all_ the above listed tasks.

* Two managers are currently being worked on: a command-line interface one and a web-interface one.

* These managers must also give the user the ability to:
    * Spawn one-shot rules with no effort (load good defaults so that rules can run with virtually no configuration).
    * Clone existing rules and use them as boilerplate for new ones.

* The web interface must also be able to provide custom interfaces for specific checks. In other words, act as `multiedit`.

#### Supervisor:

1. Loads `rules`
2. Splits them in record-based chunks
3. Sends them for processing in different nodes
4. Consults performance hints returned by the nodes to schedule their processing
5. Accumulates the record changes that have been requested by the checks and applies them to the database

#### Runner:
A runner is a processing node. It is responsible for:

* Collecting the `rules` that were sent by the supervisor
* Setting up a environment for the checks to run in, including fixtures and available records
* Notifying the user in case of errors and exceptions (via the `reporters` (see below))

#### Reporters:
A reporter is a module that allows the `runner` and the `check` to communicate with the outside world.

Reporters typically will:

* Compose and deliver a text report regarding a `rules`'s or `check`'s findings.
* Report logs, issues and exceptions.
* Spawn workflows to resolve the above.

* Reporters that are currently planned for development:

    1. Plaintext (to web and terminal)
    2. E-Mail
    3. Request Tracker
    4. Holdingpen

* Generic and specific reporters:

    1. For each `rules`, a default, generic reporter will be available. This functionality will mostly be useful for `check`s that are present in the core.
    2. On the contrary, checks that exist in overlays, may want to explicitly call their own reporters.

* Reporters may be configurable, but `check`s should ship with good defaults for them (eg: Checks that plan to use the `Holdingpen` reporter, will want to ship a custom workflow)

* Reporters must conform to a certain API, which includes:

    1. Initialization in the `supervisor` and/or `runner`.
    2. Being called from the `runner` and `check` to insert new items for reporting.
    3. A way to be informed about the status of the `rules` (ie: When the rules is completed, so that a report may be dispatched)

The above steps are key for enabling the reporters to be able to accumulate information from the nodes that belong to the same `task` run and only dispatch that information (eg with an e-mail), when the entire task has completed.

----

Please let us know if you have a use-case or requirement that the above description does not seem to cover or if you have any questions.

**Comment by aw-bib** [comment]
This sound somewhat related to the discussion in inveniosoftware/dojson#23

**Comment by kaplun** [comment]
Indeed, but not that, although jsonschema allows to describe certain type of constraints, in certain conditions one might still want to use the full power of python to perform checks. E.g.:
* checking relations across records
* checking link rotting (404s etc)
* upgrading records to a new data model
* propagating the relations towards merged records
* performing search and replace (e.g. to correct everywhere a typo in an author name)

**Comment by aw-bib** [comment]
> one might still want to use the full power of python to perform checks

Agree.

Somewhere in the discussion I just raised the point if not some such checker beyond the schema is necessary or would be good to have. E.g. in having a less strict schema that get's checked later and tagged for manual curation. Most likely one has to find some sort of _sweet spot_.
